### DISCUSSION TITLE
Class Z is implemented in both X and Y. One of the two will be used. Which one is undefined.

### DISCUSSION THREAD

**Discussion by C0rn3j** [pyinstaller/pyinstaller#8938] [Help] [answered]
How can I prevent `/opt/homebrew` from being used? I want the app to be independent of the system.

The app crashes if user has gio installed there.

CI runs with file list DEBUG logs - https://github.com/Taiko2k/Tauon/pull/1348

spec file - https://github.com/C0rn3j/Tauon/blob/4d9317eec5967d166ec3c789cfa1e2d3241c18d8/mac.spec

```python
2024-12-17 15:13:37 [ INFO  ] [  t_main  ] Rendering icons...
objc[31169]: Class GNotificationCenterDelegate is implemented in both
/Applications/TauonMusicBox.app/Contents/Frameworks/libgio-2.0.0.dylib (0x10dc8c6c0) and
/opt/homebrew/Cellar/glib/2.82.4/lib/libgio-2.0.0.dylib (0x11cd24648). 
One of the two will be used. Which one is undefined.
[1]    31169 segmentation fault  /Applications/TauonMusicBox.app/Contents/MacOS/Tauon\\ Music\\ Box
```

Similarly, I am having trouble *including* Rsvg without relying on the system version, where the only way I can get it to run is `brew install librsvg`

I already include librsvg .dyilib here https://github.com/C0rn3j/Tauon/blob/4d9317eec5967d166ec3c789cfa1e2d3241c18d8/mac.spec#L14 but it still fails:

```python
2024-12-20 11:21:40 [ INFO  ] [  t_main  ] Force rerender icons
Traceback (most recent call last):
  File "tauon/__main__.py", line 443, in <module>
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 384, in exec_module
  File "tauon/t_modules/t_main.py", line 4368, in <module>
  File "tauon/t_modules/t_main.py", line 4338, in scale_assets
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 384, in exec_module
  File "tauon/t_modules/t_svgout.py", line 27, in <module>
  File "gi/__init__.py", line 122, in require_version
ValueError: Namespace Rsvg not available
```

I presume the Rsvg issue could be related to https://github.com/orgs/pyinstaller/discussions/8147

**Comment by rokm** [comment]
That usually happens where you have a dependency that for some reason did not end up being packaged, and so the app ends up falling back to system's homebrew installation. Could be an optional dependency (or a plugin), so it will go unnoticed on systems without homebrew.

For `gi.repository.Rsvg`, you will probably need hooks to ensure that typelibs and such are collected. As far as I can see, we don't have hooks for that particular library/namespace yet.

**Reply by C0rn3j** [reply]
Is it as simple as these were?

https://github.com/pyinstaller/pyinstaller/pull/8149/files

i.e. create a hook like this (imported via `hookspath=["extra/pyinstaller-hooks"],`)

```python
#hook-gi.repository.Rsvg.py
from PyInstaller.utils.hooks.gi import GiModuleInfo

module_info = GiModuleInfo("Rsvg", "2.0")
if module_info.available:
	binaries, datas, hiddenimports = module_info.collect_typelib_data()
```

and then somehow create the safe import hook module? I am not sure where to stuff that one though.

**Reply by rokm** [reply]
> Is it as simple as these were?

It is.

The pre-safe-import hooks should be automatically discovered in the `pre_safe_import_module` sub-directory in directories passed via `hookspath`.

So if your extra hooks directory is `extra/pyinstaller-hooks`, you save the standard hook (which you've shown above) as `extra/pyinstaller-hooks/hook-gi.repository.Rsvg.py` and the pre-safe-import hook as `extra/pyinstaller-hooks/pre_safe_import_module/hook-gi.repository.Rsvg.py`.

**Reply by rokm** [reply, marked as answer]
Does collection of `Rsvg` typelib solve the issue with external `libgio-2.0.0.dylib` being loaded, or is it still being loaded from/by somewhere else? 

I suspect that falling back to system-installed `Rsvg` typelib (if it was available in Homebrew) would also load the system copy of `librsvg .dylib` and its dependencies, so this was the cause (or rather, one of the causes) of the original issue with two  `libgio-2.0.0.dylib`  instances being loaded.

**Reply by C0rn3j** [reply]
It seems to resolve it too actually!

Thanks a lot for your help!
